core/share.go

```diff
@@ -0,0 +1,57 @@
+package core
+
+import (
+	"context"
+
+	"github.com/deluan/rest"
+	gonanoid "github.com/matoous/go-nanoid"
+	"github.com/navidrome/navidrome/model"
+)
+
+type Share interface {
+	NewRepository(ctx context.Context) rest.Repository
+}
+
+func NewShare(ds model.DataStore) Share {
+	return &shareService{
+		ds: ds,
+	}
+}
+
+type shareService struct {
+	ds model.DataStore
+}
+
+func (s *shareService) NewRepository(ctx context.Context) rest.Repository {
+	repo := s.ds.Share(ctx)
+	wrapper := &shareRepositoryWrapper{
+		ShareRepository: repo,
+		Repository:      repo.(rest.Repository),
+		Persistable:     repo.(rest.Persistable),
+	}
+	return wrapper
+}
+
+type shareRepositoryWrapper struct {
+	model.ShareRepository
+	rest.Repository
+	rest.Persistable
+}
+
+func (r *shareRepositoryWrapper) Save(entity interface{}) (string, error) {
+	s := entity.(*model.Share)
+	id, err := gonanoid.Generate("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", 9)
+	s.Name = id
+	if err != nil {
+		return "", err
+	}
+	id, err = r.Persistable.Save(s)
+	return id, err
+}
+
+func (r *shareRepositoryWrapper) Update(entity interface{}, _ ...string) error {
+	s := entity.(*model.Share)
+	cols := []string{"description", "last_visited_at", "visit_count"}
```

Comment:
I don't think we should allow updating `last_visited_at` and `visit_count` from this repository. We will probably have a method like "RegisterVisit" that will update both fields when someone hits the link. This repository is for CRUD operations from the UI, and these two fields should not be editable, right?

Reply:
Can't the `registerVisit` use this update method? After all this method will probably be called from the UI (the player)?

Reply:
IMO this  `RegisterVisit` should be a method of the `Share` interface`, and it would be implemented in the `shareService`. In this way, it will be able to call the original `ShareRepository.Put` method, that allows changing any value. The Update (and all the rest.Repository's methods) is just a "connector" between the domain and the REST API, and we don't want the REST API to be able to change these values.

Navidrome's architecture is loosely based on Onion Architecture and DDD, and can be a bit confusing for new comers. Maybe it is worth for you to read a bit about these architectural patterns.

Reply:
Okay, that makes sense now. `RegisterVisit` is something of a 'private' function that can only be used by the UI, not by the public APIs, and so shouldn't be part of the repository. 
Thanks for the architectures, this was new information, this explains all the layering logic you were telling the other day.

Done the changes.

Reply:
To be clear: There are places in the code that expose fields that should not be modified to the REST API, but we are trying to avoid these issues from now on, as they are technical debt that will need to be tackled in the future